### PR TITLE
Add monotonic vertical mesh check

### PR DESCRIPTION
**Pull request by charleskawczynski**
This PR adds a check into our vertical mesh generation, to ensure that the mesh is monotonically increasing. Now that I think of it, @kmdeck, are there meshes where we want monotonically decreasing meshes? If so, we can modify the logic.

### CODE REVIEW

**Comment by szy21** [comment]
Looks good, thanks! I think Tapio wants an error message saying something like `We need to have dz_bottom <= z_max/(nz+1) and dz_top >= z_max/(nz+1)` though.

**Comment by kmdeck** [comment]
yep! for the land, we usually want finer resolution at the surface.

**Comment by kmdeck** [comment]
> yep! for the land, we usually want finer resolution at the surface.

(this can be returned to when we need it, if that is easier!)

**Comment by szy21** [comment]
> yep! for the land, we usually want finer resolution at the surface.

I think the question was if z would monotonically decrease for land, that is something like z=0 for the first grid and z<0 for the next grid?

**Comment by kmdeck** [comment]
> > yep! for the land, we usually want finer resolution at the surface.
> 
> I think the question was if z would monotonically decrease for land, that is something like z=0 for the first grid and z<0 for the next grid?

z is always defined such that it is increasing upwards. But we want our mesh finer at the surface than at the bottom, which is opposite what the atmos has, I believe.

**Comment by szy21** [comment]
> z is always defined such that it is increasing upwards. But we want our mesh finer at the surface than at the bottom, which is opposite what the atmos has, I believe.

Yep, but I think this PR is just to check if z is increasing upwards. It doesn't check what dz is doing?

**Comment by kmdeck** [comment]
> > z is always defined such that it is increasing upwards. But we want our mesh finer at the surface than at the bottom, which is opposite what the atmos has, I believe.
> 
> Yep, but I think this PR is just to check if z is increasing upwards. It doesn't check what dz is doing?

makes sense! that is consistent with the land model, then

**Comment by charleskawczynski** [comment]
Updated

File: src/Meshes/interval.jl [hunk 1]
```diff
@@ -71,6 +71,20 @@ function boundary_face_name(mesh::IntervalMesh, elem::Integer, face)
...
 
+function monotonic_check(faces)
+    n = length(faces) - 1
+    monotonic_incr = all(map(i -> faces[i].z < faces[i + 1].z, 1:n))
+    monotonic_decr = all(map(i -> faces[i].z > faces[i + 1].z, 1:n))
+    if !(monotonic_incr || monotonic_decr)
+        error(
+            string(
+                "Faces in vertical mesh are not increasing monotonically. ",
+                "We need to have dz_bottom <= z_max/(nz+1) and dz_top >= z_max/(nz+1).",
```
Comment (szy21) [inline src/Meshes/interval.jl:82]:
```suggestion
                "We need to have dz_bottom <= z_max/(z_elem+1) and dz_top >= z_max/(z_elem+1).",
```
Could you double check if this is correct (especially the +1 part)?

**Review by szy21 — APPROVED** [review]

File: src/Meshes/interval.jl (hunk 1 above)
Comment (szy21) [reply src/Meshes/interval.jl:82]: (z_elem is just to match the command line argument. I don't have a strong preference.)

Comment (charleskawczynski) [reply src/Meshes/interval.jl:82]: I'm not familiar with this relationship, @trontrytel can you confirm?

Comment (szy21) [reply src/Meshes/interval.jl:82]: Let me double check with Tapio. I actually think there should be no +1 here.

Comment (szy21) [reply src/Meshes/interval.jl:82]: (but, yes, it would be useful if @trontrytel can test!)

Comment (szy21) [reply src/Meshes/interval.jl:82]: Ok, I confirmed with Tapio, there should be no +1 here. I updated the suggestion.

**Pull request merged**